### ISSUE TITLE
Modal closes prematurely due to $locationChangeSuccess happening after modal is created

### ISSUE THREAD

**Issue by Valorum** [dwmkerr/angular-modal-service#172]
(Using Angular UI Router)

**Problem**
When I go from state A to state B (route A to route B), and in state B's controller I immediately create a modal (e.g. an error condition occurs right away, and I show an error msg using a modal), that modal will close almost immediately.

**Cause**
In the modal service, it listens for the $locationChangeSuccess event, and when that event fires, it closes the modal. This is to handle when the location changes while the modal is open, so that the modal is closed and cleaned up. This is good.

However, the problem appears to be that the $locationChangeSuccess event fires _after_ the new controller (state B's) has finished loading. This means a modal created in the controller has been created before the $locationChangeSuccess event fires. And so after the controller is done loading, and the modal has been created already, the $locationChangeSuccess event fires, which causes the model to be closed again.

**Solution**
My current work-around is to wrap the showModal() call in a $timeout() with 0 duration. This puts the showModal call on a new tick, and gives the location change time to finish up on the current tick. This nicely wraps up all events associated with the route change first.

A possible solution would be to wrap the setup of the listener for the $locationChangeSucces inside a $timeout, e.g. change

```
var rootScopeOnClose = $rootScope.$on('$locationChangeSuccess', cleanUpClose);
```

to

```
var rootScopeOnClose;
$timeout(function() {
    rootScopeOnClose = $rootScope.$on('$locationChangeSuccess', cleanUpClose);
});
```

**Comment by eraye1** [comment]
Ran into this issue too.  👍.

**Comment by pdemilly** [comment]
Same problem here. I am try to use it in an onEnter: clause of ui-router but the modal promise resolves as soon as you open it which completes the onEnter promise which generates a $locationChangeSuccess which closes the modal

**Comment by alejandroiglesias** [comment]
I also ran into this issue and fixed it by wrapping the `ModalService.showModal` call in a `$timeout`.

**Comment by dwmkerr** [comment]
Resolved in `0.13` with the `locationChangeSuccess` option: https://github.com/dwmkerr/angular-modal-service#showmodal-options
